File: use-cases/easy-ship/code/java/src/main/java/lambda/InventoryCheckHandler.java
package lambda;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.swagger.client.model.easyship.UnitOfLength;
import io.swagger.client.model.easyship.UnitOfWeight;
import io.swagger.client.model.easyship.Weight;
import io.swagger.client.model.easyship.Dimensions;
import lambda.utils.*;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.google.gson.Gson;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.GetItemRequest;
import software.amazon.awssdk.services.dynamodb.model.GetItemResponse;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

import static java.lang.Long.parseLong;
import static lambda.utils.Constants.*;

public class InventoryCheckHandler implements RequestHandler<StateMachineInput, StateMachineInput> {

    @Override
    public StateMachineInput handleRequest(StateMachineInput input, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("InventoryCheck Lambda input: " + new Gson().toJson(input));

        long packageWeightValue = 0;
        String packageWeightUnit = "";
        long packageLength = 0;
        long packageWidth = 0;
        long packageHeight = 0;
        String packageSizeUnit = "";

        for (EasyShipOrderItem orderItem : input.getEasyShipOrder().getOrderItems()) {
            //Retrieve the item from DynamoDB by SKU
            //Update this section to match your product's logic
            Map<String, AttributeValue> key = new HashMap<>();
            key.put(INVENTORY_TABLE_HASH_KEY_NAME, AttributeValue.fromS(orderItem.getSku()));
            GetItemRequest getItemRequest = GetItemRequest.builder()
                    .tableName(System.getenv(INVENTORY_TABLE_NAME_ENV_VARIABLE))
                    .key(key)
                    .build();

            DynamoDbClient dynamoDB = DynamoDbClient.builder().build();
            GetItemResponse getItemResult = dynamoDB.getItem(getItemRequest);
            Map<String, AttributeValue> item = getItemResult.item();

            String stock = item.get(INVENTORY_TABLE_STOCK_ATTRIBUTE_NAME).n();

            if (parseLong(stock) < orderItem.getQuantity()) {
                throw new InternalError(
                        String.format("Stock level for SKU {%s} is not enough to fulfill the requested quantity",
                                orderItem.getSku()));
            }

            long itemWeightValue = parseLong(item.get(INVENTORY_TABLE_WEIGHT_VALUE_ATTRIBUTE_NAME).n());
            //Valid values for the database records are uppercase: [OZ, G]
            String itemWeightUnit = item.get(INVENTORY_TABLE_WEIGHT_UNIT_ATTRIBUTE_NAME).s();

            long itemLength = parseLong(item.get(INVENTORY_TABLE_LENGTH_ATTRIBUTE_NAME).n());
            long itemWidth = parseLong(item.get(INVENTORY_TABLE_WIDTH_ATTRIBUTE_NAME).n());
            long itemHeight = parseLong(item.get(INVENTORY_TABLE_HEIGHT_ATTRIBUTE_NAME).n());
            //Valid values for the database records are uppercase: [INCHES, CENTIMETERS]
            String itemSizeUnit = item.get(INVENTORY_TABLE_SIZE_UNIT_ATTRIBUTE_NAME).s();

            Weight itemWeight = new Weight();
            itemWeight.setValue(BigDecimal.valueOf(itemWeightValue));
            itemWeight.setUnit(UnitOfWeight.valueOf(itemWeightUnit));

            //Package weight is calculated by adding the individual weights
            //Update this section to match your selling partners' logic
            packageWeightValue += itemWeightValue;
            packageWeightUnit = itemWeightUnit;

            //Package size is calculated by adding the individual sizes
            //Update this section to match your selling partners' logic
            packageLength += itemLength;
            packageWidth += itemWidth;
            packageHeight += itemHeight;
            packageSizeUnit = itemSizeUnit;
        }

        input.getEasyShipOrder().setPackageWeight(new Weight()
                .value(BigDecimal.valueOf(packageWeightValue))
                .unit(UnitOfWeight.valueOf(packageWeightUnit)));

        input.getEasyShipOrder().setPackageDimensions(new Dimensions()
                .length(BigDecimal.valueOf(packageLength))
                .width(BigDecimal.valueOf(packageWidth))
                .height(BigDecimal.valueOf(packageHeight))
                .unit(UnitOfLength.valueOf(packageSizeUnit)));

        return input;
    }
}

File: use-cases/easy-ship/code/php/lambda/InventoryCheckHandler.php
<?php

namespace Lambda;

use Aws\DynamoDb\DynamoDbClient;
use Aws\DynamoDb\Exception\DynamoDbException;
use Lambda\Utils\Constants;
use Lambda\Utils\LambdaContext;
use Lambda\Utils\Model\StateMachineInput;
use Lambda\Utils\StateMachineInputConverter;
use SpApi\Model\easyship\v2022_03_23\Dimensions;
use SpApi\Model\easyship\v2022_03_23\Weight;

class InventoryCheckHandler
{
    /**
     * @throws \Exception
     */
    public function handleRequest(array $event, LambdaContext $context): StateMachineInput
    {
        $input = StateMachineInputConverter::convertFromArray($event);
        $logger = $context->getLogger();
        $logger->info('InventoryCheck Lambda input: ' . json_encode($event, JSON_PRETTY_PRINT));

        $packageWeightValue = 0;
        $packageWeightUnit = '';
        $packageLength = 0;
        $packageWidth = 0;
        $packageHeight = 0;
        $packageSizeUnit = '';

        $dynamoDb = new DynamoDbClient([
            'region' => getenv('AWS_REGION'),
            'version' => 'latest'
        ]);

        foreach ($input->getEasyShipOrder()->getOrderItems() as $orderItem) {
            // Retrieve the item from DynamoDB by SKU
            $key = [
                Constants::INVENTORY_TABLE_HASH_KEY_NAME => ['S' => $orderItem->getSku()]
            ];

            try {
                $result = $dynamoDb->getItem([
                    'TableName' => getenv(Constants::INVENTORY_TABLE_NAME_ENV_VARIABLE),
                    'Key' => $key
                ]);

                $item = $result['Item'];

                if ($item === null) {
                    $logger->error(sprintf("Item not found for SKU: %s", $orderItem->getSku()));
                    throw new \Exception(sprintf(
                        "Item not found for SKU {%s} in DynamoDB",
                        $orderItem->getSku()
                    ));
                }

                $stock = $item[Constants::INVENTORY_TABLE_STOCK_ATTRIBUTE_NAME]['N'];
                if ((int)$stock < $orderItem->getQuantity()) {
                    throw new \Exception(sprintf(
                        "Stock level for SKU {%s} is not enough to fulfill the requested quantity",
                        $orderItem['sku']
                    ));
                }

                $itemWeightValue = (int)$item[Constants::INVENTORY_TABLE_WEIGHT_VALUE_ATTRIBUTE_NAME]['N'];
                $itemWeightUnit = $item[Constants::INVENTORY_TABLE_WEIGHT_UNIT_ATTRIBUTE_NAME]['S'];

                $itemLength = (int)$item[Constants::INVENTORY_TABLE_LENGTH_ATTRIBUTE_NAME]['N'];
                $itemWidth = (int)$item[Constants::INVENTORY_TABLE_WIDTH_ATTRIBUTE_NAME]['N'];
                $itemHeight = (int)$item[Constants::INVENTORY_TABLE_HEIGHT_ATTRIBUTE_NAME]['N'];
                $itemSizeUnit = $item[Constants::INVENTORY_TABLE_SIZE_UNIT_ATTRIBUTE_NAME]['S'];

                // Package weight is calculated by adding the individual weights
                $packageWeightValue += $itemWeightValue;
                $packageWeightUnit = $itemWeightUnit;

                // Package size is calculated by adding the individual sizes
                $packageLength += $itemLength;
                $packageWidth += $itemWidth;
                $packageHeight += $itemHeight;
                $packageSizeUnit = $itemSizeUnit;
            } catch (DynamoDbException $e) {
                $logger->error("Unable to get item: " . $e->getMessage());
                throw $e;
            }
        }

        $packageWeight = new Weight([
            'value' => $packageWeightValue,
            'unit' => $packageWeightUnit
        ]);

        $input->getEasyShipOrder()->setPackageWeight($packageWeight);

        $packageDimensions = new Dimensions([
            'length' => $packageLength,
            'width' => $packageWidth,
            'height' => $packageHeight,
            'unit' => $packageSizeUnit
        ]);

        $input->getEasyShipOrder()->setPackageDimensions($packageDimensions);

        return $input;
    }
}
